docs/plugins.md

```diff
@@ -0,0 +1,138 @@
+# Creating an Argo Rollouts Plugin
+
+## High Level Overview
+
+Argo Rollouts plugins depend on hashicorp's [go-plugin](https://github.com/hashicorp/go-plugin) library. This library 
+provides a way for a plugin to be compiled as a standalone executable and then loaded by the rollouts controller at runtime.
+This works by having the plugin executable act as a rpc server and the rollouts controller act as a client. The plugin executable
+is started by the rollouts controller and is a long-lived process and that the rollouts controller connects to over a unix socket.
+The communication protocol uses golang built in net/rpc library so plugins have to be written in golang.
+
+## Plugin Repository
+
+In order to get plugins listed in the main argo rollouts documentation we ask that the plugin repository be created under
+the [argoproj-labs](https://github.com/argoproj-labs) organization. Please open an issue under argo-rollouts requesting a 
+repo which you would be granted admin access on. 
+
+There is also a standard naming convention for plugin names used for configmap registration, as well as what the plugin 
+uses for locating its specific configuration on rollout or analysis resources. The name needs to be in the form of 
+`<namespace>/<name>` and both <namespace> and <name> have a regular expression check that matches Github's requirements 
+for `username/org` and `repository name`. This requirement is in place to help with allowing multiple creators of the same plugin 
+types to exist such as `<org1>/nginx` and `<org2>/nginx`. These names could be based of the repo name such 
+as `argoproj-labs/rollouts-sample_prometheus-metric-plugin` but it is not a requirement. 
+
+There will also be a standard for naming repositories under argoproj-labs in the form of `rollouts-<tool>-<type>-plugin`
+where `<type>` is say `metric`, or `trafficrouter` and `<tool>` is the software the plugin is for say nginx.
+
+## Plugin Name
+
+So now that we have an idea on plugin naming and repository standards let's pick a name to use for the rest of this 
+documentation and call our plugin `argoproj-labs/nginx`.
+
+This name will be used in a few different spots the first is the config map that your plugin users will need to configure.
+It looks like this below.
+
+```yaml
+kind: ConfigMap
+metadata:
+  name: argo-rollouts-config
+data:
+  plugins: |-
```

Comment:
This pattern of providing configurations brings problems when we want to have some level of composition in how the application is defined in different environments. For example, if we want to use kustomize to define a `trafficRouter` to be used in all environments but want to enable different `metricProviders` in just a few, we are stuck. I would suggest to at least separate `metricProviders` from `trafficRouters` configurations.